server/homebrew.api.spec.js

```diff
@@ -0,0 +1,377 @@
+			views       : 1,
+			lastViewed  : new Date(),
+			version     : 1,
+			pageCount   : 1,
+			textBin     : ''
+		};
+		googleBrew = {
+			...hbBrew,
+			googleId : '12345'
+		};
+	});
+
+	afterEach(()=>{
+		jest.restoreAllMocks();
+	});
+
+	describe('getId', ()=>{
+		it('should return only id if google id is not present', ()=>{
+			const { id, googleId } = api.getId({
+				params : {
+					id : 'abcdefgh'
+				}
+			});
+
+			expect(id).toEqual('abcdefgh');
+			expect(googleId).toBeUndefined();
+		});
+
+		it('should return id and google id from request body', ()=>{
+			const { id, googleId } = api.getId({
+				params : {
+					id : 'abcdefgh'
+				},
+				body : {
+					googleId : '12345'
+				}
+			});
+
+			expect(id).toEqual('abcdefgh');
+			expect(googleId).toEqual('12345');
+		});
+
+		it('should return id and google id params', ()=>{
+			const { id, googleId } = api.getId({
+				params : {
+					id : '123456789012abcdefghijkl'
+				}
+			});
+
+			expect(id).toEqual('abcdefghijkl');
+			expect(googleId).toEqual('123456789012');
+		});
+	});
+
+	describe('getBrew', ()=>{
+		const toBrewPromise = (brew)=>new Promise((res)=>res({ toObject: ()=>brew }));
+		const notFoundError = 'Brew not found in Homebrewery database or Google Drive';
+
+		it('returns middleware', ()=>{
+			const getFn = api.getBrew('share');
+			expect(getFn).toBeInstanceOf(Function);
+		});
+
+		it('should fetch from mongoose', async ()=>{
+			const testBrew = { title: 'test brew', authors: [] };
+			api.getId = jest.fn(()=>({ id: '1', googleId: undefined }));
+			model.get = jest.fn(()=>toBrewPromise(testBrew));
+
+			const fn = api.getBrew('share', true);
+			const req = { brew: {} };
+			const next = jest.fn();
+			await fn(req, null, next);
+
+			expect(req.brew).toEqual(testBrew);
+			expect(next).toHaveBeenCalled();
+			expect(api.getId).toHaveBeenCalledWith(req);
+			expect(model.get).toHaveBeenCalledWith({ shareId: '1' });
+		});
+
+		it('should handle mongoose error', async ()=>{
+			api.getId = jest.fn(()=>({ id: '1', googleId: undefined }));
+			model.get = jest.fn(()=>new Promise((_, rej)=>rej('Unable to find brew')));
+
+			const fn = api.getBrew('share', true);
+			const req = { brew: {} };
+			const next = jest.fn();
+			let err;
+			try {
+				await fn(req, null, next);
+			} catch (e) {
+				err = e;
+			}
+
+			expect(err).toEqual(notFoundError);
+			expect(req.brew).toEqual({});
+			expect(next).not.toHaveBeenCalled();
+			expect(api.getId).toHaveBeenCalledWith(req);
+			expect(model.get).toHaveBeenCalledWith({ shareId: '1' });
+		});
+
+		it('changes tags from string to array', async ()=>{
+			const testBrew = { title: 'test brew', authors: [], tags: 'tag' };
+			api.getId = jest.fn(()=>({ id: '1', googleId: undefined }));
+			model.get = jest.fn(()=>toBrewPromise(testBrew));
+
+			const fn = api.getBrew('share', true);
+			const req = { brew: {} };
+			const next = jest.fn();
+			await fn(req, null, next);
+
+			expect(req.brew.tags).toEqual([]);
```

Comment:
Just out of my own curiosity, could you please explain why we see an empty array as `tags` here? What exactly are we testing here? Do I understand it correctly that we may have string representation for tags in our database and we have some code which drops that field and replaces it with an empty array? If so, why do we need that special handling, can't we ensure that DB already contains data in a correct format?

Reply:
You are correct that there are string representations for tags in the database.

Doing a mass-conversion to an empty array would be a high-impact database operation, possibly resulting in some downtime. We don't have formal procedures for "downtime for maintenance", so such an operation has been punted. For now.

Additionally, the code does need to handle brews with invalid values .. which might be retrieved from google drive storage. Somewhere there's a bit of work with implementing a Postel Robustness filter on the retrieve/save functions. IIRC this work hasn't been refactored in.

Reply:
To clarify:

Old brews have strings for tags, but *they are all empty strings*. We switched to arrays before the Tags UI was implemented, so there was no way to actually create a tag when it was still strings. That is, all actual tags are correctly implemented in arrays, new brews will have an empty array, and it is only old brews that may have an empty string.

Reply:
Yep! This test is merely testing the logic that exists in the getBrew function that performs the replacement from string to empty array.